cpp/examples/arrow/dataset-documentation-example.cc

```diff
@@ -217,24 +229,29 @@ std::shared_ptr<arrow::Table> SelectAndProjectDataset(
   auto scan_builder = dataset->NewScan().ValueOrDie();
   std::vector<std::string> names;
   std::vector<ds::Expression> exprs;
+  // Read all the original columns.
   for (const auto& field : dataset->schema()->fields()) {
     names.push_back(field->name());
     exprs.push_back(ds::field_ref(field->name()));
   }
+  // Also derive a new column.
   names.push_back("b_large");
   exprs.push_back(ds::greater(ds::field_ref("b"), ds::literal(1)));
   ABORT_ON_FAILURE(scan_builder->Project(exprs, names));
```

Comment:
From what I've seen, we've tried to do this directly on the ScannerBuilder:

```cpp
    std::vector<std::string>   projection_attrs;
    arrow::dataset::Expression selection_expr;

    ...

    ARROW_RETURN_NOT_OK(scanbuilder->Project(projection_attrs));
    ARROW_RETURN_NOT_OK(scanbuilder->Filter(selection_expr));

    ...
```

I am looking at the code first, so I may just need to read the actual documentation included in this PR, but it may be nice to include a comment that references the documentation of why to prefer one approach to another or if they're equivalent mechanisms.

Reply:
Ah, the example here is using Project to define new virtual columns from the physical columns, while your snippet is simply selecting a subset of the physical columns to read. The former API is a superset of the latter. There's an example of what you currently do a little bit prior to this in the file. I've added explanatory comments for these examples, though in the prose I think it should also be clear.